### ISSUE TITLE
Underfitting model with `@parametrised` 

### ISSUE THREAD

**Issue by patel-zeel** [wesselb/stheno#19]
### Description of the bug

Hi @wesselb,

I am not sure if this issue is better suited in [varz](https://github.com/wesselb/varz/tree/master/varz). I found an abnormal behaviour in the optimization process when using two equivalent variants of defining the model.

* Variant 1: As per #18 - Predictive variance seems properly calibrated in final fit.
![image](https://user-images.githubusercontent.com/59758528/142761082-5d050bdc-a665-449e-9d93-5e32f4e5518e.png)

* Variant 2: Enabling `@parametrised` version of code in #18 - Predictive variance seems underfitting in final fit.
![image](https://user-images.githubusercontent.com/59758528/142761122-3f978681-4903-49e7-993e-bf5417db7a68.png)

I tried disabling fixed random seed and still experienced the same phenomenon.

### Description of your environment

Package                       Version              Location
----------------------------- -------------------- -----------------------
absl-py                       0.13.0
alabaster                     0.7.12
algebra                       1.1.0
argon2-cffi                   21.1.0
astunparse                    1.6.3
attrs                         21.2.0
autograd                      1.3
Babel                         2.9.1
backcall                      0.2.0
backends                      1.4.13
backends-matrix               1.2.6
black                         21.9b0
bleach                        4.1.0
cachetools                    4.2.2
certifi                       2021.5.30
cffi                          1.14.6
cftime                        1.5.1
charset-normalizer            2.0.6
clang                         5.0
click                         8.0.2
cloudpickle                   2.0.0
coverage                      5.5
coveralls                     3.2.0
cycler                        0.10.0
debugpy                       1.5.0
decorator                     5.1.0
defusedxml                    0.7.1
dm-tree                       0.1.6
docopt                        0.6.2
docutils                      0.17.1
entrypoints                   0.3
fdm                           0.4.1
flatbuffers                   1.12
future                        0.18.2
gast                          0.4.0
google-auth                   1.32.0
google-auth-oauthlib          0.4.4
google-pasta                  0.2.0
gpytorch                      1.5.1
grpcio                        1.34.1
h5py                          3.1.0
idna                          3.2
imagesize                     1.2.0
iniconfig                     1.1.1
ipykernel                     6.4.1
ipython                       7.28.0
ipython-genutils              0.2.0
ipywidgets                    7.6.5
jax                           0.2.22
jaxlib                        0.1.72
jedi                          0.18.0
Jinja2                        3.0.2
joblib                        1.1.0
jsonschema                    4.1.0
jupyter                       1.0.0
jupyter-client                7.0.6
jupyter-console               6.4.0
jupyter-core                  4.8.1
jupyterlab-pygments           0.1.2
jupyterlab-widgets            1.0.2
keras                         2.6.0
keras-nightly                 2.5.0.dev2021032900
Keras-Preprocessing           1.1.2
kiwisolver                    1.3.2
Markdown                      3.3.4
MarkupSafe                    2.0.1
matplotlib                    3.4.3
matplotlib-inline             0.1.3
mistune                       0.8.4
mlkernels                     0.3.4
mpmath                        1.2.1
mypy-extensions               0.4.3
nbclient                      0.5.4
nbconvert                     6.2.0
nbformat                      5.1.3
nest-asyncio                  1.5.1
netCDF4                       1.5.7
notebook                      6.4.4
numpy                         1.21.4
oauthlib                      3.1.1
opt-einsum                    3.3.0
packaging                     21.0
pandas                        1.3.3
pandocfilters                 1.5.0
parso                         0.8.2
pathspec                      0.9.0
pexpect                       4.8.0
pickleshare                   0.7.5
Pillow                        8.3.2
pip                           21.2.4
platformdirs                  2.4.0
pluggy                        1.0.0
plum-dispatch                 1.5.4
prometheus-client             0.11.0
prompt-toolkit                3.0.20
protobuf                      3.17.3
ptyprocess                    0.7.0
py                            1.10.0
pyasn1                        0.4.8
pyasn1-modules                0.2.8
pycparser                     2.20
Pygments                      2.10.0
pyparsing                     2.4.7
pyrsistent                    0.18.0
pytest                        6.2.5
pytest-cov                    3.0.0
python-dateutil               2.8.2
python-slugify                5.0.2
pytz                          2021.3
pyzmq                         22.3.0
qtconsole                     5.1.1
QtPy                          1.11.2
redeal                        0.2.0
regdata                       1.0.2
regex                         2021.10.8
requests                      2.26.0
requests-oauthlib             1.3.0
rsa                           4.7.2
scikit-learn                  1.0
scipy                         1.7.1
Send2Trash                    1.8.0
setuptools                    58.0.4
setuptools-scm                6.3.2
setuptools-scm-git-archive    1.1
six                           1.15.0
sklearn                       0.0
snowballstemmer               2.1.0
Sphinx                        4.2.0
sphinx-rtd-theme              1.0.0
sphinxcontrib-applehelp       1.0.2
sphinxcontrib-devhelp         1.0.2
sphinxcontrib-htmlhelp        2.0.0
sphinxcontrib-jsmath          1.0.1
sphinxcontrib-qthelp          1.0.3
sphinxcontrib-serializinghtml 1.1.5
stheno                        1.1.8.dev21+gaa3ce24 /home/patel_zeel/stheno
sympy                         1.9
tensorboard                   2.6.0
tensorboard-data-server       0.6.1
tensorboard-plugin-wit        1.8.0
tensorflow                    2.5.0
tensorflow-estimator          2.5.0
tensorflow-probability        0.13.0
termcolor                     1.1.0
terminado                     0.12.1
testpath                      0.5.0
text-unidecode                1.3
threadpoolctl                 3.0.0
toml                          0.10.2
tomli                         1.2.1
torch                         1.9.1
tornado                       6.1
traitlets                     5.1.0
typing_extensions             4.0.0
urllib3                       1.26.7
varz                          0.7.4
wbml                          0.3.14
wcwidth                       0.2.5
webencodings                  0.5.1
Werkzeug                      2.0.2
wheel                         0.37.0
widgetsnbextension            3.5.1
wrapt                         1.12.1
xarray                        0.19.0

**Comment by wesselb** [comment]
Hey @patel-zeel,

Thanks for your other PR with the example and for opening this issue. :)

What's going wrong is that, in the prediction, the noise variance is not added (so only the uncertainty over the function is shown):

```python
...

def plot_model(vs, title_prefix):
    f, noise = model(vs)
    f_post = f | (f(x_obs, noise), y_obs)
    mean, lower, upper = f_post(x).marginal_credible_bounds()

    ...
```

If you also want to plot the noise variance, then you can do this as follows:

```python
...

def plot_model(vs, title_prefix):
    f, noise = model(vs)
    f_post = f | (f(x_obs, noise), y_obs)
    # `noise` is added below:      vvvvv 
    mean, lower, upper = f_post(x, noise).marginal_credible_bounds()

    ...
```

The reason why variant 1 is not showing the pathological behaviour is because, in that case, the noise variance is not learned and set to something higher than the true noise variance (I'll elaborate below), which means that the functional uncertainty also will be higher.

In

```python
...
y_obs = f_true + noise_true * B.randn(x_obs.shape[0])
...
```

`noise_true` should have been `noise_true ** 0.5`. Currently, the noise that is actually added is `noise_true ** 2 = 0.01` rather than `noise_true = 0.1`, which is much lower.

I hope this clarifies what's going on!

**Comment by patel-zeel** [comment]
Thank you for clarifying with a detailed answer as always @wesselb :) I completely missed the fact that `noise` is learned in `@parametrised` version. I have updated #18 incorporating these changes.
